### PR TITLE
fix prompt strip to support tensors and np arrays

### PR DESCRIPTION
**Pull request by AvivSham**
# What does this PR do?

`WhisperTokenizer` does not strip the prompt ids when calling `decode` for `torch.Tensor`, `tf.Tensor`, and `np.ndarray`. It does not perform the slicing in `_strip_prompt` because the condition `isinstance(token_ids, list)` is not met.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@sanchit-gandhi
@ArthurZucker

### CODE REVIEW

**Comment by AvivSham** [comment]
I was not able to make it pass the CI/CD tests due to import issues. @sanchit-gandhi can you please guide?

File: src/transformers/models/whisper/tokenization_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/tokenization_whisper.py:856]: Let's convert the token ids to a list **before** checking whether we have a prompt. If we convert the token ids to a list first, then we don't need to check:
```python
isinstance(token_ids, (List[int], np.ndarray, torch.Tensor, tf.Tensor))
```
=> since we know `token_ids` will be a list!

File: src/transformers/models/whisper/tokenization_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/tokenization_whisper.py:866]: What you can do here is check whether `token_ids` has the method `numpy`, and subsequently convert it to a numpy array:
```python
if hasattr(token_ids, "numpy"):
    token_ids = token_ids.numpy()
# now the token ids are either a numpy array, or a list of lists
if isinstance(token_ids, np.ndarray):
    token_ids = token_ids.tolist()
# now the token ids are always a list of lists
return token_ids
```

It might be more robust checking that we perform this on Torch/TF tensors:
```python
if "torch" in str(type(token_ids)) or "tensorflow" in str(type(token_ids)) and hasattr(token_ids, "numpy"):
    token_ids = token_ids.numpy()
# now the token ids are either a numpy array, or a list of lists
if isinstance(token_ids, np.ndarray):
    token_ids = token_ids.tolist()
# now the token ids are always a list of lists
return token_ids
```

**Review by sanchit-gandhi — COMMENTED** [review]
Hey @AvivSham! Thanks for opening this PR - the first draft looks in decent nick!

One thing we need to change is how we check for Torch/TF tensors: the feature extraction file should be _agnostic_ to framework. Imagine we have a user running Whisper in PyTorch. We do not force them to have TF installed, since there's no need to have both PyTorch and TF installed if just using PyTorch. Therefore, we cannot assume TF is an available import, and thus can't leverage it in the feature extraction file. The converse is true for someone using TF, where we don't expect them to have PyTorch installed.

=> what we need to do here is refactor the changes such that they are _agnostic_ to framework. I've left some hints on how to do this below! Let me know if you have any questions - happy to help!

File: src/transformers/models/whisper/tokenization_whisper.py
Comment (AvivSham) [reply src/transformers/models/whisper/tokenization_whisper.py]: the first option is more elegant than the second but I decided to take the second.

**Comment by AvivSham** [comment]
Thank you for reviewing @sanchit-gandhi !
I was not aware of the framework agnostic constraint, I think we are ok now, can you please review it again?

**Comment by AvivSham** [comment]
This PR is not stale, waiting for @sanchit-gandhi to approve.

**Comment by amyeroberts** [comment]
Gentle ping @sanchit-gandhi

**Comment by AvivSham** [comment]
@amyeroberts @sanchit-gandhi 
Any news about this PR?

**Comment by AvivSham** [comment]
@amyeroberts This issue is not stale, still waiting for @sanchit-gandhi.

**Comment by amyeroberts** [comment]
@sanchit-gandhi @ylacombe Can one of you review this?

File: src/transformers/models/whisper/tokenization_whisper.py
Comment (sanchit-gandhi) [reply src/transformers/models/whisper/tokenization_whisper.py]: This is still valid! We need to convert to list **before** checking whether we have a prompt

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks for the last round of commits @AvivSham! Super sorry for the late review here

I would recommend using the following script to quickly debug your changes and confirm they're working (currently the following code fails since we don't strip the prompt for `numpy` inputs):

```python
import numpy as np
from transformers import WhisperTokenizer

prompt_text = "the cat sat on the mat"
input_text = "the quick brown fox"

tokenizer = WhisperTokenizer.from_pretrained("openai/whisper-tiny")
prompt_ids = tokenizer.get_prompt_ids("the cat sat on the mat", return_tensors="np")

input_ids = tokenizer("the quick brown fox", return_tensors="np").input_ids[0]

input_ids = np.hstack([prompt_ids, input_ids])

# check with prompt in output
pred_text = tokenizer.decode(input_ids, skip_special_tokens=False)
assert pred_text.strip() == prompt_text + input_text

# check stripping prompt from output
pred_text = tokenizer.decode(input_ids, skip_special_tokens=True)
assert pred_text.strip() == input_text
```

Along with the subtle changes to the order of operations (which will fix the above example), it would be great to add a slow/fast tokenizer test to confirm we get the correct behaviour after this change. To do this, you can add a test in the file `test_tokenization_whisper.py`. I would follow the existing logic to load a pre-trained slow and fast (or "rust") tokenizers, then run a similar check to the one we have above. You can use the following test as an example:
https://github.com/huggingface/transformers/blob/ce3647ad2d2f94b0858aa0145f6de9eb677327b9/tests/models/whisper/test_tokenization_whisper.py#L269-L271

Let me know if you have any questions! Happy to help with the final changes before merge!

**Comment by AvivSham** [comment]
Hi @sanchit-gandhi,
Added the requested changes (including adding the test), can you review again (hopefully for the last time 😄)?

File: src/transformers/models/whisper/tokenization_whisper_fast.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/tokenization_whisper_fast.py:585]:
```suggestion
    # Copied from transformers.models.whisper.tokenization_whisper.WhisperTokenizer._strip_prompt
    def _strip_prompt(self, token_ids: List[int], prompt_token_id: int, decoder_start_token_id: int):
```

File: src/transformers/models/whisper/tokenization_whisper_fast.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/tokenization_whisper_fast.py:598]: Let's copy this method from `tokenization_whisper.py`, as suggested [here](https://github.com/huggingface/transformers/pull/27818/files#r1634779349):

```suggestion
    @staticmethod
    # Copied from transformers.models.whisper.tokenization_whisper.WhisperTokenizer. _convert_to_list
    def _convert_to_list(token_ids):
```

File: tests/models/whisper/test_tokenization_whisper.py
Comment (sanchit-gandhi) [inline tests/models/whisper/test_tokenization_whisper.py:109]: Has this `expected_encoding` stayed the same? I can't spot any differences between the two dicts. If so, can we revert the changes? It would be super helpful to keep the PR isolated to the changes required to fix the prompt strip

File: tests/models/whisper/test_tokenization_whisper.py
Comment (sanchit-gandhi) [inline tests/models/whisper/test_tokenization_whisper.py:141]: Likewise here - if `next_sequences_1` is unchanged, can we revert this change?

File: tests/models/whisper/test_tokenization_whisper.py
Comment (sanchit-gandhi) [inline tests/models/whisper/test_tokenization_whisper.py:268]: Beautiful test 👌

**Review by sanchit-gandhi — COMMENTED** [review]
Hey @AvivSham! The functionality looks great here! Just a few clean-up comments below. Once done, could you run:
```
make style
```
In order to run the linter to format the code? If you hit issues regarding missing packages, you can install them with the following from the root of the Transformers repo:
```
pip install -e ".[quality]"
```
Many thanks!

File: src/transformers/models/whisper/tokenization_whisper_fast.py
Comment (AvivSham) [reply src/transformers/models/whisper/tokenization_whisper_fast.py:585]: ohh I thought you wanted to remove it, that your intention was to put this comment for my info.

File: tests/models/whisper/test_tokenization_whisper.py
Comment (AvivSham) [reply tests/models/whisper/test_tokenization_whisper.py:109]: I guess that this is from the `transformers` main branch, I will revert it to before the pull

**Comment by AvivSham** [comment]
@sanchit-gandhi 
Done all the changes required including reverting and running the style. Please check again.

File: src/transformers/models/whisper/tokenization_whisper_fast.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/tokenization_whisper_fast.py:599]:
```suggestion
    # Copied from transformers.models.whisper.tokenization_whisper.WhisperTokenizer._convert_to_list
```

**Review by sanchit-gandhi — COMMENTED** [review]
Looks like there was a minor typo (suggested a fix below), which running `make fix-copies` should check is correct. By the way, there's great documentation on the style checks we use in Transformers in this [contributing guide](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request).

It looks like a Whisper tokenizer test is failing on the CI for this PR ([traceback](https://app.circleci.com/pipelines/github/huggingface/transformers/95601/workflows/81baf434-1668-4fe1-a3b6-9bd00a008c56/jobs/1259000)). You can run the test locally from the root of the repo with the following command:
```sh
pytest -sv tests/models/whisper/test_tokenization_whisper.py::WhisperTokenizerTest::test_skip_special_tokens_with_timestamps
```
Could you have a look to see why this test is failing after these PR changes? Happy to help if it's unclear! If you have any dependency issues when running the test, you can install the necessary packages with:
```sh
pip install -e ".[quality]"
```

**Comment by AvivSham** [comment]
Hi @sanchit-gandhi ,
This test fails since `token_ids` is an empty list. IMO this should be handled elsewhere and much earlier maybe in [this line](https://github.com/huggingface/transformers/pull/27818/files#diff-b258211cb184b5aebf1a50b0f4a3f63683823996d367b68ac5afc504353d5c21R644), I cannot this about a case that user will try to decode empty list. Maybe `_preprocess_token_ids` would be a better place to add it.
WDYT?  
if you agree I can add this to the current PR:
```python
if not isinstance(token_ids, int):
    if len(token_ids) == 0:
        raise ValueError("No token ids provided for decoding.")
```
or if you see a reason for decoding an empty string we can return `token_ids` instead of raising the exception.

**Comment by sanchit-gandhi** [comment]
Hey @AvivSham - sorry I don't fully follow! If we look at the test [test_skip_special_tokens_with_timestamps](https://github.com/huggingface/transformers/blob/1e79eade41c346d1d8c2ad79c9e32e3186f20b54/tests/models/whisper/test_tokenization_whisper.py#L209), we see that we always pass a non-empty array of token-ids to the tokeniser's `.decode` method. Running this test on `main`, we see that we always get the expected decoding results. For this PR, we need to make sure the changes we implement do not change the current decoding behaviour, which means that the test should pass. I would recommend using the debugger to check the behaviour of `_preprocess_token_ids` and `_strip_prompt` before and after the changes, and try to unify this PR to how it is originally done. If you get stuck, let me know and I can take a deeper look!

**Comment by AvivSham** [comment]
Ok, so I debugged it and the empty `token_ids` comes from the method `_decode_with_timestamps` check line `553` under `tokenization_whisper.py` file.
I made a small adjustment to address this issue, and all the tests passed successfully.
@sanchit-gandhi please have another look at it 😄

**Comment by AvivSham** [comment]
@amyeroberts @sanchit-gandhi

**Comment by amyeroberts** [comment]
Hi @AvivSham, you need to run `make fix-copies` to resolve the failing quality checks. 

cc @kamalkraj For first review as @sanchit-gandhi is off

**Comment by AvivSham** [comment]
@amyeroberts I ran it and committed.

**Comment by AvivSham** [comment]
@amyeroberts @kamilakesbi 
Gentle reminder 😃

**Review by kamilakesbi — COMMENTED** [review]
Hi @AvivSham, thanks for working on this! 

It LGTM! we can merge this PR @amyeroberts :)

File: src/transformers/models/whisper/tokenization_whisper.py
Comment (amyeroberts) [reply src/transformers/models/whisper/tokenization_whisper.py:873]: This comment appears not to be resolved. @AvivSham could you provide details on why it wasn't implemented or add the suggestion?

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! Just one comment

File: src/transformers/models/whisper/tokenization_whisper.py
Comment (AvivSham) [reply src/transformers/models/whisper/tokenization_whisper.py:873]: Hi @amyeroberts,
This is a mistake we can resolve this comment. See [these lines](https://github.com/huggingface/transformers/pull/27818/files#diff-2fb1a089a6a7c4bbbab1eba43e5bfe7fdd29415e24fc62c1ea8601ea14f9ad48R606-R613).

**Comment by AvivSham** [comment]
Hi @amyeroberts,
This is a mistake we can resolve this comment. See [these lines](https://github.com/huggingface/transformers/pull/27818/files#diff-2fb1a089a6a7c4bbbab1eba43e5bfe7fdd29415e24fc62c1ea8601ea14f9ad48R606-R613).

**Review by amyeroberts — APPROVED** [review]
Thanks for adding!

**Comment by AvivSham** [comment]
@amyeroberts I keep receiving emails regarding CI/CD failed runs how can we prevent that? I thought it would stop once the PR is merged.
image

**Comment by amyeroberts** [comment]
Hi @AvivSham, this is because of the settings on your fork. 

You can disable the CI runs by going to the repo, then Settings > Actions > Disable actions

image

**Comment by AvivSham** [comment]
@amyeroberts Thank you very much for helping.

**Pull request merged**
